### PR TITLE
WIP: GaussianSource based on new libpyvinyl

### PR DESCRIPTION
**Pull request by CFGrote**

### CODE REVIEW

File: SimExLite/SourceCalculators/GaussianSourceCalculator.py [hunk 1]
```diff
@@ -0,0 +1,207 @@
+""":module GaussianSourceCalculator: Module that holds the GaussianSourceCalculator class.  """
+
+import sys
+import numpy as np
+from pathlib import Path
+from scipy.constants import hbar, c
+from pint import Quantity, Unit
+import imp
+
```
Comment (CFGrote) [inline SimExLite/SourceCalculators/GaussianSourceCalculator.py:9]: General: imports should be sorted alphabetically, `from .. import ..` go after plain `import ..`s. Check if imported modules are unused.

File: SimExLite/SourceCalculators/GaussianSourceCalculator.py [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+from libpyvinyl import BaseCalculator, CalculatorParameters
+from SimExLite.WavefrontData import WavefrontData
+
+
+class GaussianSourceCalculator(BaseCalculator):
+    """:class GaussianSourceCalculator: Class representing a x-ray free electron laser photon source."""
```
Comment (CFGrote) [inline SimExLite/SourceCalculators/GaussianSourceCalculator.py:15]: docstring: mention "Gaussian approximation"

File: SimExLite/SourceCalculators/GaussianSourceCalculator.py [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+
+    def __init__(
+        self,
+        name: str,
+        WPG_path: str,
```
Comment (CFGrote) [inline SimExLite/SourceCalculators/GaussianSourceCalculator.py:20]: remove `WPG_path`: wpg must be installed in the environment or be on $PYTHONPATH.

File: SimExLite/SourceCalculators/GaussianSourceCalculator.py [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+        pulse_energy = parameters.new_parameter(
+            "pulse_energy", comment="Total energy of the pulse", unit="joule"
+        )
+        pulse_energy.value = 2e-3
+
+        divergence = parameters.new_parameter(
+            "divergence", comment="Beam divergence angle", unit="radian"
+        )
+        divergence.add_interval(0, 2 * np.pi, True)
+        diver_value = get_divergence_from_beam_diameter(
```
Comment (CFGrote) [inline SimExLite/SourceCalculators/GaussianSourceCalculator.py:89]: `divergence_value = ...` and remove line 92

File: SimExLite/SourceCalculators/GaussianSourceCalculator.py [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+        num_time_slices = parameters.new_parameter(
+            "number_of_time_slices", comment="The number of time slices"
+        )
+        num_time_slices.add_interval(0, None, True)
+        num_time_slices.value = 12
+
+        param_z = parameters.new_parameter(
+            "z", comment="The position of the pulse in the beam direction", unit="m"
+        )
+        param_z.value = 100
```
Comment (CFGrote) [inline SimExLite/SourceCalculators/GaussianSourceCalculator.py:116]: The z position should be a parameter of the calculator, ie. the user can set it.

File: SimExLite/SourceCalculators/GaussianSourceCalculator.py [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+
+        self.parameters = parameters
+
+    def backengine(self):
+        from wpg.generators import build_gauss_wavefront
```
Comment (CFGrote) [inline SimExLite/SourceCalculators/GaussianSourceCalculator.py:121]: why local imports?

File: SimExLite/SourceCalculators/GaussianSourceCalculator.py [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+            / self.parameters["photon_energy_relative_bandwidth"].value
+            / E_joule
+        )
+        pulse_energy = self.parameters["pulse_energy"].value_no_conversion
+
+        beam_waist = 2.0 * hbar * c / theta / E_joule
+        wavelength = 1239.8e-9 / E_eV
+        rayleigh_length = np.pi * beam_waist**2 / wavelength
+
+        print("rayleigh_length =", rayleigh_length)
```
Comment (CFGrote) [inline SimExLite/SourceCalculators/GaussianSourceCalculator.py:143]: remove print, use logging if needed

File: SimExLite/SourceCalculators/GaussianSourceCalculator.py [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+        # Store on class.
+        srwl_wf.Rx = Rx
+        srwl_wf.Ry = Ry
+
+        key = self.output_keys[0]
+        filename = self.output_filenames[0]
+        output_data = self.output[key]
+
+        wavefront = Wavefront(srwl_wf)
+        wavefront.store_hdf5(filename)
```
Comment (CFGrote) [inline SimExLite/SourceCalculators/GaussianSourceCalculator.py:193]: why is IO part of the backengine? i thought the new way of handling data is that backengine returns a Data object and IO is handled somewhere else or by the caller of the Calculator instance?

File: SimExLite/WavefrontData/WavefrontData.py [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+            key,
+            filename=filename,
+            file_format_class=format_class,
+            file_format_kwargs=kwargs,
+        )
+
+    @classmethod
+    def from_dict(cls, data_dict, key):
+        """Create the data class by a python dictionary."""
+        return cls(key, data_dict=data_dict)
```
Comment (CFGrote) [inline SimExLite/WavefrontData/WavefrontData.py:57]: super! I really like our new style of handling data!

File: tests/test_GaussianSourceCalculator.py [hunk 10]
```diff
@@ -0,0 +1,19 @@
+"""Test EMCPhoton"""
+
+import pytest
+from SimExLite.SourceCalculators import GaussianSourceCalculator
+
+WPG_path = "/home/juncheng/GPFS/exfel/data/user/juncheng/WPG"
+
+
```
Comment (CFGrote) [inline tests/test_GaussianSourceCalculator.py:8]: more detailled testing needed:
* test construction with and without default parameters
* test each method
* test a full construct-setup-run-collect workflow
* test dump/load to/from disk

File: tests/test_WavefrontData.py [hunk 11]
```diff
@@ -0,0 +1 @@
+from SimExLite.WavefrontData import WavefrontData
```
Comment (CFGrote) [inline tests/test_WavefrontData.py:1]: guess this needs to be completed but is  very important to test properly

**Review by CFGrote — COMMENTED** [review]
good job juncheng! i added my review below.

File: SimExLite/SourceCalculators/GaussianSourceCalculator.py (hunk 8 above)
Comment (JunCEEE) [reply SimExLite/SourceCalculators/GaussianSourceCalculator.py:193]: Yes the backengine is returning the Data object, and here the Data object is a "file mapping" (line 195). The lines you are referring to is the file produced by the backengine. Maybe what you want is a "python object mapping", but it's not the case here. 
 
https://github.com/PaNOSC-ViNYL/SimEx-Lite/blob/40ec02a7f69ee1e5f545c05e844fc7bfcb1d6536/SimExLite/SourceCalculators/GaussianSourceCalculator.py#L192-L195

File: SimExLite/SourceCalculators/GaussianSourceCalculator.py (hunk 3 above)
Comment (JunCEEE) [reply SimExLite/SourceCalculators/GaussianSourceCalculator.py:20]: Unfortunately, I didn't find an instruction to install WPG into the environment in their documentation. Also, it's not possible to `export PYTHONPATH=XXX` in a jupyter-hub environment. 

Do you know any good way to install WPG into the environment or set the $PYTHONPATH in a jupyter-hub environment?

Comment (CFGrote) [reply SimExLite/SourceCalculators/GaussianSourceCalculator.py:20]: I'd write somewhere in the documentation:
SimEx-Lite depends on WPG (github.com/Samoyl/WPG). Please download and build WPG as per their documentation. Then, you have to make the WPG library "visible" to SimEx-Lite. To this end, do one of the following:
1. `$PYTHONPATH`: If you have the environment variable `$PYTHONPATH` set, copy the WPG/ directory into any directory listed in `$PYTHONPATH`.
2. `conda`: If you installed SimEx-Lite in a conda environment, activate that environment and copy the WPG/ directory to $CONDA_PREFIX/lib/python3.x/site-packages/ (x is your python minor version).
3. `pip`: If you installed SimEx-Lite with `pip install`, copy the WPG directory into $HOME/.local/lib/python3.x/site-packages

Test that WPG can be imported by running:
```console
$> python -c 'import WPG.Beamline'
```

Comment (CFGrote) [reply SimExLite/SourceCalculators/GaussianSourceCalculator.py:20]: Needless to say, these instructions should be tested and corrected if needed

File: SimExLite/SourceCalculators/GaussianSourceCalculator.py (hunk 5 above)
Comment (CFGrote) [reply SimExLite/SourceCalculators/GaussianSourceCalculator.py:116]: no thats totally fine. but here the z value is set by the calculator itself. i'm saying z should be coming from the code that constructs/configures the calculator

File: SimExLite/SourceCalculators/GaussianSourceCalculator.py (hunk 1 above)
Comment (JunCEEE) [reply SimExLite/SourceCalculators/GaussianSourceCalculator.py:10]: Besides to sort them alphabetically, I would like to have the global package first and the local package after.

File: SimExLite/SourceCalculators/GaussianSourceCalculator.py (hunk 2 above)
Comment (JunCEEE) [reply SimExLite/SourceCalculators/GaussianSourceCalculator.py:24]: Done.

File: SimExLite/SourceCalculators/GaussianSourceCalculator.py (hunk 3 above)
Comment (JunCEEE) [reply SimExLite/SourceCalculators/GaussianSourceCalculator.py]: This problem will be fixed in this issue: #15

Comment (JunCEEE) [reply SimExLite/SourceCalculators/GaussianSourceCalculator.py]: This problem will be fixed in this issue: #15

File: SimExLite/SourceCalculators/GaussianSourceCalculator.py (hunk 5 above)
Comment (JunCEEE) [reply SimExLite/SourceCalculators/GaussianSourceCalculator.py:108]: Do you mean something like:
`calculator = GaussianSourceCalculator("gaussian_source", parameters=parameters, z=100)`?
I would think that "z" one of the `parameters`. Wouldn't that be a bit confusing?

File: tests/test_GaussianSourceCalculator.py (hunk 10 above)
Comment (JunCEEE) [reply tests/test_GaussianSourceCalculator.py:7]: We are dealing with them in this PR: https://github.com/PaNOSC-ViNYL/SimEx-Lite/pull/14

File: tests/test_WavefrontData.py (hunk 11 above)
Comment (JunCEEE) [reply tests/test_WavefrontData.py:1]: It will be addressed in this issue: https://github.com/PaNOSC-ViNYL/SimEx-Lite/issues/16

**Comment by JunCEEE** [comment]
Waiting for the final feedback from @CFGrote . After that, this can be merged (rebase).

**Pull request merged**